src/Sentry/Internal/SentryScopeManager.cs

```diff
@@ -10,65 +10,69 @@ namespace Sentry.Internal
             set => _asyncLocalScope.Value = value;
         }
 
-        private Func<ImmutableStack<ValueTuple<Scope, ISentryClient>>> NewStack { get; }
+        private Func<Stack<KeyValuePair<Scope, ISentryClient>>> NewStack { get; }
 
         public SentryScopeManager(
             SentryOptions options,
             ISentryClient rootClient)
         {
             Debug.Assert(rootClient != null);
             _options = options;
-            NewStack = () => ImmutableStack.Create(new ValueTuple<Scope, ISentryClient>(new Scope(options), rootClient));
+            NewStack = () => new Stack<KeyValuePair<Scope, ISentryClient>>(new [] { new KeyValuePair<Scope, ISentryClient>(new Scope(options), rootClient) });
         }
 
-        public ValueTuple<Scope, ISentryClient> GetCurrent() => ScopeAndClientStack.Peek();
+        public KeyValuePair<Scope, ISentryClient> GetCurrent() => ScopeAndClientStack.Peek();
 
         public void ConfigureScope(Action<Scope> configureScope)
         {
             _options?.DiagnosticLogger?.LogDebug("Configuring the scope.");
             var scope = GetCurrent();
-            configureScope?.Invoke(scope.Item1);
+            configureScope?.Invoke(scope.Key);
         }
 
         public Task ConfigureScopeAsync(Func<Scope, Task> configureScope)
         {
             _options?.DiagnosticLogger?.LogDebug("Configuring the scope asynchronously.");
             var scope = GetCurrent();
-            return configureScope?.Invoke(scope.Item1) ?? Task.CompletedTask;
+            return configureScope?.Invoke(scope.Key) ?? Task.CompletedTask;
         }
 
         public IDisposable PushScope() => PushScope<object>(null);
 
         public IDisposable PushScope<TState>(TState state)
         {
             var currentScopeAndClientStack = ScopeAndClientStack;
-            var (scope, client) = currentScopeAndClientStack.Peek();
+            var scope = currentScopeAndClientStack.Peek();
 
-            if (scope.Locked)
+            if (scope.Key.Locked)
             {
                 // TODO: keep state on current scope?
                 _options?.DiagnosticLogger?.LogDebug("Locked scope. No new scope pushed.");
                 return DisabledHub.Instance;
             }
 
-            var clonedScope = scope.Clone();
+            var clonedScope = scope.Key.Clone();
 
             if (state != null)
             {
                 clonedScope.Apply(state);
             }
+
             var scopeSnapshot = new ScopeSnapshot(_options, currentScopeAndClientStack, this);
             _options?.DiagnosticLogger?.LogDebug("New scope pushed.");
-            ScopeAndClientStack = currentScopeAndClientStack.Push(new ValueTuple<Scope, ISentryClient>(clonedScope, client));
-
+            var newScopeAndClientStack = new Stack<KeyValuePair<Scope, ISentryClient>>(currentScopeAndClientStack.Count + 1);
+            foreach (var item in currentScopeAndClientStack)
```

Comment:
currentScopeAndClientStack is now a copy to a mutable reference so this would throw if we're iterating over it while it changes

Reply:
Not sure I understand. `currentScopeAndClientStack` is never modified. newScopeAndClientStack is a copy off `currentScopeAndClientStack`

Reply:
It's just: `var currentScopeAndClientStack = ScopeAndClientStack;` so it's just a reference to `ScopeAndClientStack` which can be mutated by anyone calling `SentrySdk.PushScope` for example.

When it was immutable, we didn't have to worry. Now we need make it thread-safe (`ConcurrentStack<T>` would likely do) otherwise the collection can be mutated while iterating over the items in order to copy them.

Reply:
Still don't understand. No one can modify `ScopeAndClientStack`-property because it is internal. And those calling `SentrySdk.PushScope` makes a clone of `currentScopeAndClientStack` and assigns a whole new object to `ScopeAndClientStack`-property. Think you have to show some code that can make your scenario happen.

Reply:
On a second pass it doesn't look like we're mutating the list anywhere after it's creation so the older behavior is here.

It is racey though if two threads called `PushScope` concurrently. It was already actually while using Immutable collections. since we copy the reference to make a clone, then assign to the property to be made a side effect. So looks like it needs a refactor.

Unrelated to your change. So I'll resolve this issue and I'll do another pass later on the IDE itself. Sorry the delay

Reply:
`ScopeAndClientStack` uses AsyncLocal (Advanced Thread-Static with async task support), so you have a references for each thread-execution-context. The modifications from one thread will not affect thread-execution-context of other concurrent threads.

Reply:
Yes that's true. This is one thing the .NET SDK diverges with all the other Sentry SDKs. After having 2 other SDKs crashing because of concurrency issues on the Hub I even forgot we had AsyncLocal here. 2 years is a long time.